go/kbfs/simplefs/archive.go

```diff
@@ -0,0 +1,876 @@
+
+	manifest := make(map[string]keybase1.SimpleFSArchiveFile)
+	gotList := false
+loopReadList:
+	for {
+		listResult, err := m.simpleFS.SimpleFSReadList(ctx, opid)
+		if err != nil || len(listResult.Entries) == 0 {
+			if gotList {
+				break loopReadList
+			}
+		}
+		gotList = true
+
+		for _, e := range listResult.Entries {
+			manifest[e.Name] = keybase1.SimpleFSArchiveFile{
+				State:      keybase1.SimpleFSFileArchiveState_ToDo,
+				DirentType: e.DirentType,
+			}
+		}
+	}
+
+	func() {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+
+		jobCopy, ok := m.state.Jobs[jobID]
+		if !ok {
+			m.simpleFS.log.CWarningf(ctx, "job %s not found. it might have been canceled", jobID)
+			return
+		}
+		jobCopy.Manifest = manifest
+		m.state.Jobs[jobID] = jobCopy
+	}()
+	return nil
+}
+
+func (m *archiveManager) indexingWorker(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-m.indexingWorkerSignal:
+		}
+
+		jobID, jobCtx, ok := m.startWorkerTask(ctx,
+			keybase1.SimpleFSArchiveJobPhase_Queued,
+			keybase1.SimpleFSArchiveJobPhase_Indexing)
+
+		if !ok {
+			continue
+		}
+		// We got a task. Put another token into the signal channel so we
+		// check again on the next iteration.
+		m.signal(m.indexingWorkerSignal)
+
+		m.simpleFS.log.CDebugf(ctx, "indexing: %s", jobID)
+
+		err := m.doIndexing(jobCtx, jobID)
+		switch err {
+		case nil:
+			m.simpleFS.log.CDebugf(jobCtx, "indexing done on job %s", jobID)
+			m.changeJobPhase(jobCtx, jobID, keybase1.SimpleFSArchiveJobPhase_Indexed)
+		default:
+			m.simpleFS.log.CErrorf(jobCtx, "indexing error on job %s: %v", jobID, err)
+			m.setJobError(ctx, jobID, err)
+		}
+
+		m.fluseStateFile(ctx)
+		m.signal(m.copyingWorkerSignal) // Done copying! Notify the copying worker.
+	}
+}
+
+type sha256TeeReader struct {
+	inner          io.Reader
+	innerTeeReader io.Reader
+	h              hash.Hash
+}
+
+var _ io.Reader = (*sha256TeeReader)(nil)
+
+// Read implements the io.Reader interface.
+func (r *sha256TeeReader) Read(p []byte) (n int, err error) {
+	return r.innerTeeReader.Read(p)
+}
+
+func (r *sha256TeeReader) getSum() []byte {
+	return r.h.Sum(nil)
+}
+
+func newSHA256TeeReader(inner io.Reader) (r *sha256TeeReader) {
+	r = &sha256TeeReader{
+		inner: inner,
+		h:     sha256.New(),
+	}
+	r.innerTeeReader = io.TeeReader(r.inner, r.h)
+	return r
+}
+
+func (m *archiveManager) copyFileFromBeginning(ctx context.Context,
+	srcDirFS billy.Filesystem, entryPathWithinJob string,
+	localPath string, mode os.FileMode) (sha256Sum []byte, err error) {
+	m.simpleFS.log.CDebugf(ctx, "+ copyFileFromBeginning %s", entryPathWithinJob)
+	defer func() { m.simpleFS.log.CDebugf(ctx, "- copyFileFromBeginning %s err: %v", entryPathWithinJob, err) }()
+
+	src, err := srcDirFS.Open(entryPathWithinJob)
+	if err != nil {
+		return nil, fmt.Errorf("srcDirFS.Open(%s) error: %v", entryPathWithinJob, err)
+	}
+	defer src.Close()
+
+	dst, err := os.OpenFile(localPath, os.O_APPEND|os.O_WRONLY|os.O_CREATE, mode)
```

Comment:
You don't want `O_APPEND` if you're copying from the beginning of the file, right?  Seems like this wouldn't worth in the "recover after corruption detection" case in the function below this.

Reply:
Good catch! I'm also realizing I didn't close that file (which is opened for sha256sum) before starting over so fixing that as well